decoder.go

```diff
@@ -74,42 +45,23 @@ func frameLineSize(frame *C.AVFrame) *C.int {
 	return (*C.int)(unsafe.Pointer(&frame.linesize[0]))
 }
 
-// getStreamInfo opens a stream URL and retrieves the video codec.
-func getStreamInfo(url string) (videoCodec, error) {
-	cUrl := C.CString(url)
-	defer C.free(unsafe.Pointer(cUrl))
+// getAvailableCodec returns a supported codec in the given tracks.
+func getAvailableCodec(tracks media.Medias) videoCodec {
+	var h264 *formats.H264
+	var h265 *formats.H265
 
-	var avFormatCtx *C.AVFormatContext = nil
-	ret := C.avformat_open_input(&avFormatCtx, cUrl, nil, nil)
-	if ret < 0 {
-		return Unknown, fmt.Errorf("avformat_open_input() failed: %s", avError(ret))
-	}
-	defer C.avformat_close_input(&avFormatCtx)
-
-	ret = C.avformat_find_stream_info(avFormatCtx, nil)
-	if ret < 0 {
-		return Unknown, fmt.Errorf("avformat_find_stream_info() failed: %s", avError(ret))
-	}
+	// List of formats/codecs in priority order
+	formatPointers := []interface{}{&h264, &h265}
```

Comment:
this should be a slice of `format.Format` not interface{} see https://github.com/erh/viamrtsp/pull/7/files#diff-616e82b0c6eddaf46a49b2e6fa5146effd4c74559a0f0f152df3d330d6f4d9e4R35

Reply:
Yes, I was trying to use `format.Format` here but was running into issues.

- The library shows using a pointer to the format and then memory ref which is essentially a double pointer...
    - The double pointer no longer satisfies the interface
```
./rtsp.go:421:37: cannot use &h264 (value of type **formats.H264) as type formats.Format in array or slice literal:
        **formats.H264 does not implement formats.Format (missing ClockRate method)
./rtsp.go:421:44: cannot use &h265 (value of type **formats.H265) as type formats.Format in array or slice literal:
        **formats.H265 does not implement formats.Format (missing ClockRate method)
make: *** [Makefile:24: bin/viamrtsp] Error 2
```
- If I just use a single pointer, the interface is satisfied. However, the `findFormat` fails (does not pick up any formats).
- I also tried creating new format instances using `new` without success. (no format is found).

Reply:
Ah, I see you are upgraded to `v4` in that PR. https://github.com/erh/viamrtsp/pull/7/files#diff-616e82b0c6eddaf46a49b2e6fa5146effd4c74559a0f0f152df3d330d6f4d9e4R10

Reply:
I have not been able to successfully use `format.Format`.

`FindFormat` takes in an empty interface:
```
// FindFormat finds a certain format among all the formats in all the medias of the stream.
// If the format is found, it is inserted into forma, and its media is returned.
func (d *Session) FindFormat(forma interface{}) *Media {
	for _, media := range d.Medias {
		ok := media.FindFormat(forma)
		if ok {
			return media
		}
	}
	return nil
}
```

This aligns with demo code that passes in the memory address of the pointer to the format (which does not satisfy the Format interface).